### PR TITLE
[IMP] hr_payroll: towards multi contract support

### PR DESCRIPTION
**Pull request by RomainLibert**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/hr_contract/models/hr_contract.py [hunk 1]
```diff
@@ -109,6 +104,27 @@ def _onchange_employee_id(self):
...
+        """ Two contracts in state [incoming | pending | open] cannot overlap """
+        for contract in self.filtered(lambda c: c.state not in ['draft', 'close', 'cancel']):
+            domain = [
+                ('id', '!=', contract.id),
+                ('employee_id', '=', contract.employee_id.id),
+                ('state', 'in', ['incoming', 'pending', 'open']),
+            ]
+
+            if not contract.date_end:
+                start_domain = [('date_start', '>=', contract.date_start)]
```
Comment (RomainLibert) [inline addons/hr_contract/models/hr_contract.py]: In this case I think the final domain should look like this `['|', ('date_end', '>=', contract.date_start), ('date_end', '=', False)]`
Which means 
```
start_domain = []
end_domain = ['|', ('date_end', '>=', contract.date_start), ('date_end', '=', False)]
```

File: addons/hr_payroll/models/hr_leave.py [hunk 2]
```diff
@@ -13,6 +16,36 @@ class HrLeaveType(models.Model):
...
 
+    @api.multi
+    def _create_resource_leave(self):
+        """
+            We override this method to add the leave in the calendars of contracts.
+            Its needed when generating payslips to have leave hours corresponding
+            to the working hours of contracts.
+        """
+        super(HrLeave, self)._create_resource_leave()
+        values = []
```
Comment (RomainLibert) [inline addons/hr_payroll/models/hr_leave.py]: Careful, a leave might not have an employee linked to it

ps : I find something like this clearer
```
values = []

for leave in self:
    if leave.employee_id:
        vals = {
          'name': leave.name,
          'date_from': leave.date_from,
          'holiday_id': leave.id,
          'date_to': leave.date_to,
          'resource_id': leave.employee_id.resource_id.id,
          'time_type': leave.holiday_status_id.time_type
        }

        calendars = leave.employee_id._get_contracts(leave.date_from, leave.date_to).mapped('resource_calendar_id')
        values += [dict(vals, calendar_id=calendar.id) for calendar in calendars]

self.env['resource.calendar.leaves'].create(values)
```

File: addons/hr_payroll/models/hr_leave.py [hunk 3]
```diff
@@ -71,3 +129,33 @@ def action_refuse(self):
...
+        super(HrLeave, self)._get_number_of_days(date_from, date_to, employee_id)
+        if employee_id:
+            employee = self.env['hr.employee'].browse(employee_id)
+            # Use sudo otherwise base users can't compute number of days
+            contracts = employee.sudo()._get_contracts(date_from, date_to, states=['incoming', 'open', 'pending'])
+            calendar = contracts[:1].resource_calendar_id if contracts else None # Note: if len(contracts)>1, the leave creation will crash because of unicity constaint
+            return employee.get_work_days_data(date_from, date_to, calendar=calendar)['days']
+
+        time_delta = date_to - date_from
+        return math.ceil(time_delta.days + float(time_delta.seconds) / 86400)
```
Comment (RomainLibert) [inline addons/hr_payroll/models/hr_leave.py]: if no employee return the result from super call -> no code duplication

File: addons/hr_payroll/models/hr_leave.py [hunk 4]
```diff
@@ -71,3 +129,33 @@ def action_refuse(self):
...
+
+    @api.multi
+    @api.depends('number_of_days')
+    def _compute_number_of_hours_display(self):
```
Comment (RomainLibert) [inline addons/hr_payroll/models/hr_leave.py]: Check that the code is still correct see fix (https://github.com/odoo/odoo/commit/765075ff475e0120d1c3f5564f2259c34d2386cc)

File: addons/hr_payroll/models/hr_payslip.py [hunk 5]
```diff
@@ -154,16 +150,17 @@ def get_worked_day_lines(self, contracts, date_from, date_to):
         @return: returns a list of dict containing the worked days that should be applied for the given contract between date_from and date_to
         """
         res = []
+        # TODO remove for loop ? we only have one contract per payslip
```
Comment (RomainLibert) [inline addons/hr_payroll/models/hr_payslip.py]: probably

File: addons/hr_payroll/models/resource.py [hunk 6]
```diff
@@ -16,6 +17,24 @@ class ResourceCalendar(models.Model):
         'resource.calendar.attendance', 'calendar_id', 'Employees working Time',
         domain=[('resource_id', '!=', False)])
 
+    @api.multi
+    def transfer_leaves_to(self, other_calendar, resource=None, from_date=None):
```
Comment (RomainLibert) [inline addons/hr_payroll/models/resource.py]: rename argument `resource` as it is not obvious that you can pass multiple resources

File: addons/hr_contract/models/hr_contract.py [hunk 7]
```diff
@@ -109,6 +104,27 @@ def _onchange_employee_id(self):
...
+            if not contract.date_end:
+                start_domain = []
+                end_domain = ['|', ('date_end', '>=', contract.date_start), ('date_end', '=', False)]
+            else:
+                start_domain = [('date_start', '<=', contract.date_end)]
+                end_domain = ['|', ('date_end', '>', contract.date_start), ('date_end', '=', False)]
+
+            domain = expression.AND([domain, start_domain, end_domain])
+            if self.search_count(domain):
+                raise ValidationError(_('An employee can only have one contract incoming, open or to renew at the same time.'))
```
Comment (sswapnesh) [inline addons/hr_contract/models/hr_contract.py]: I think we should improve error message 
`An employee can have only one contract in Incoming, open or to renew state at the same time.`
or 
`An employee can have only one Running contract at the same time.
`

File: addons/hr_contract/models/hr_contract.py [hunk 8]
```diff
@@ -16,23 +16,17 @@ class Employee(models.Model):
     medic_exam = fields.Date(string='Medical Examination Date', groups="hr.group_hr_user")
     vehicle = fields.Char(string='Company Vehicle', groups="hr.group_hr_user")
     contract_ids = fields.One2many('hr.contract', 'employee_id', string='Employee Contracts')
-    contract_id = fields.Many2one('hr.contract', compute='_compute_contract_id', string='Current Contract', help='Latest contract of the employee')
+    contract_id = fields.Many2one('hr.contract', string='Current Contract', help='Latest contract of the employee')
```
Comment (tivisse) [inline addons/hr_contract/models/hr_contract.py]: Check to keep a compute method.

Comment (Musvol) [reply addons/hr_contract/models/hr_contract.py]: When the employee has a new (open) contract, his calendar becomes the one from the contract.
And if a CDI is ended to begin another contract, calendar.leaves should be transferred to the new contract's calendar.
It's easier to detect a contract change if it's not computed.

**Comment by tivisse** [comment]
@robodoo r+

**Pull request closed without merging**